### ISSUE TITLE
files are written as expected, but the files are empty, size of all files is 0 bytes

### ISSUE THREAD

**Issue by resession** [mscdex/busboy#277]
here is the function that i am using

```js
const { Readable } = require('stream')
handleFormData(folderPath, headers, data){
  const bb = busboy({ headers })
  return new Promise((resolve, reject) => {
    bb.on('file', (name, file, info) => {
      file.pipe(fs.createWriteStream(path.join(folderPath, info.filename)))
    })
    bb.on('error', error => {
      reject(error)
    })
    bb.on('close', () => {
      resolve(null)
    })
    Readable.from(data).pipe(bb)
  })
}
```

i have tried the following from the front-end

```js
for(const data of arrayOfFiles){
  FormData.append('file', data)
}
```

```js
for(const data of arrayOfFiles){
FormData.append('file', new Blob([await data.text()]))
// inside an async function
}
// this one one at least writes the files to disk but cuts off at a few KB, it does not write the files fully, the test files are over 5 MB.`
```

the `Content-Type` header looks like `multipart/form-data; boundary=----WebKitFormBoundaryHRAJ7MozJorr9xAJ`, i don't specify the header myself, i let the browser handle the headers that way it is done correctly.

nothing works.

tl;dr files are successfully written to disk, but all of the files are empty(0 bytes)

any ideas or fixes that come to mind for this problem?

**Comment by mscdex** [comment]
Have you verified the contents of the stream you're piping to `busboy` looks correct?

On an unrelated note, you should not use `info.filename` as-is for security reasons.

**Comment by resession** [comment]
> Have you verified the contents of the stream you're piping to `busboy` looks correct?
> 
> On an unrelated note, you should not use `info.filename` as-is for security reasons.

i had nodejs v16. downgraded to v14(didn't help), then upgraded to v16 again.

when i do formdata.append('file', data),i looked at the contents being piped to busboy, the content for some reason only is showing the headers and the actual file contents are not being shown, and then after a few headers, the error event fires. nothing is written to disk(understandable).

when i do formdata.append('file', new Blob([await data.text()])), the contents seem right. it shows the headers and it shows the actual content of the files. the problem is nothing is written to disk, the close event never fires and it hangs, it never finishes.

**Comment by mscdex** [comment]
If you can provide the raw request data that reproduces the problem (you can verify this by saving the raw request data to disk and then `fs.createReadStream('rawrequest').pipe(bb)`), I can take a look.

It might also be helpful to try with a normal HTML form to see if the problem still exists. If it does, then there might be a busboy usage issue.

**Comment by resession** [comment]
i tried doing `new FormData(document.getElementById('someformid'))`
i saved the raw request in a file, and this is what i got in the file
`------WebKitFormBoundary9ZEpiBS7uRCcLLi4--`
that's it. nothing was written to file, and the error event fired. this is the error.
`Error: Unexpected end of form`

then i tried `var someFormData = new FormData()
      for(const data of this.someFiles){
        // console.log(data)
        someFormData.append('file', data, data.name)
      }`
saved the request to a file, this is what i got
`------WebKitFormBoundarywA2lwuAxqwWcp3O6
Content-Disposition: form-data; name="file"; filename="about.ce8b39e4.js"
Content-Type: text/javascript


------WebKitFormBoundarywA2lwuAxqwWcp3O6
Content-Disposition: form-data; name="file"; filename="about.ce8b39e4.js.map"
Content-Type: text/plain


------WebKitFormBoundarywA2lwuAxqwWcp3O6
Content-Disposition: form-data; name="file"; filename="app.85893be4.js"
Content-Type: text/javascript


------WebKitFormBoundarywA2lwuAxqwWcp3O6
Content-Disposition: form-data; name="file"; filename="app.85893be4.js.map"
Content-Type: text/plain


------WebKitFormBoundarywA2lwuAxqwWcp3O6
Content-Disposition: form-data; name="file"; filename="app.a6f0cac2.css"
Content-Type: text/css


------WebKitFormBoundarywA2lwuAxqwWcp3O6
Content-Disposition: form-data; name="file"; filename="chunk-vendors.05d965ab.js"
Content-Type: text/javascript


------WebKitFormBoundarywA2lwuAxqwWcp3O6
Content-Disposition: form-data; name="file"; filename="chunk-vendors.05d965ab.js.map"
Content-Type: text/plain


------WebKitFormBoundarywA2lwuAxqwWcp3O6
Content-Disposition: form-data; name="file"; filename="chunk-vendors.8bda0bd5.css"
Content-Type: text/css


------WebKitFormBoundarywA2lwuAxqwWcp3O6
Content-Disposition: form-data; name="file"; filename="favicon.ico"
Content-Type: image/x-icon


------WebKitFormBoundarywA2lwuAxqwWcp3O6
Content-Disposition: form-data; name="file"; filename="index.html"
Content-Type: text/html


------WebKitFormBoundarywA2lwuAxqwWcp3O6--
`
nothing was written to disk, error event fires.
`Error: Unexpected end of form`

as you can see, i get no file contents, only the headers.

some background, i am making a fetch api for webtorrent using make-fetch https://github.com/RangerMauve/make-fetch
make-fetch takes the body and turns it into an async-iterator. that is why in my first message here in the first code block, you see `Readable.from(data).pipe(bb)` inside the promise.

i wanted to test it against something. so i made a quick express app with multer(which uses busboy). i used the same front-end fetch request code with the express/multer app.
`new FormData(document.getElementById('someformid'))` does not work.
BUT `var someFormData = new FormData()
      for(const data of this.someFiles){
        // console.log(data)
        someFormData.append('file', data, data.name)
      }` works as it should. files are written to disk and the request gets the correct response.

so i think there is a possibility that make-fetch is changing the incoming request data which then throws off busboy which might have a hard time parsing things.

**Comment by mscdex** [comment]
There's nothing I can do if `busboy` isn't receiving any form fields or any form field data. The "Unexpected end of form" typically means the boundary in the Content-Type header doesn't match what was given in the request body or the body was empty or some other format.

It sounds like you'll need to create an issue on the issue tracker of the project you're using to generate or submit the form on the client side?

**Comment by resession** [comment]
using multiple projects/lib so gotta get to the bottom of where its going wrong. looks like busboy is not the issue. thank you for checking it out with me. i'll go ahead and close out the issue.
